Import gxt

lexnewgate opened the pull request:
#22 
I have integrated the none-ui part. You can test it by uncomment the following code in `GXTWindow.cs`.
```csharp
 protected override void OnWindowOpened()
        {
            base.OnWindowOpened();
            //foreach (var kv in GTX.Gtx.EntryNameWordDict)
            //{
	           // Debug.LogError($"k {kv.Key} v {kv.Value}");
            //}
        }
``` 
I am intertested in UI part. However it may take some time for me. So I think I'd better submit the none-ui part first. 

There are some thoughts I commented in code. It may look messy. I will delete those after I put them in issues next time.

Commit:
import GTX code

import GTX code
add some empty function before integrate

Commit:
set gtx path

Commit:
refactor code

in0finite commented on Assets/Scripts/Behaviours/Loader.cs:
```diff
@@ -99,6 +100,7 @@ private static void AddLoadingSteps ()
 				new LoadingStep ( StepSetSplash2, "Set splash 2" ),
 				new LoadingStep ( StepLoadMap, "Loading map", 2.1f ),
 				new LoadingStep ( StepLoadSpecialTextures, "Loading special textures", 0.01f ),
+				new LoadingStep(StepLoadGTX ,"Loading GTX",0.1f) //todo will calculate time  
```
typo: GXT

in0finite commented on Assets/Scripts/Importing/GTX/GTX.cs:
```diff
@@ -0,0 +1,242 @@
+
+namespace SanAndreasUnity.Importing.GXT
+{
+    public class GTX : IDisposable
```
typo: GXT

in0finite commented on Assets/Scripts/Importing/GTX/GTX.cs:
```diff
@@ -0,0 +1,242 @@
+
+	    #region LoadPath
+		//this whole region needs refactoring
+	    enum Language
```
it may be better to enumerate files in "text" directory, but it's ok

lexnewgate replied:
Check #24 
It may be better because someone may not have a certain language gxt file? Do I understand you right?

in0finite commented on Assets/Scripts/UI/GXTWindow.cs:
```diff
@@ -0,0 +1,59 @@
+        GXTWindow()
+        {
+            //this will use as menu item name as well as window title
+            this.windowName = "GTXWindow";
```
typo: GXT

in0finite commented:
I think you didn't understand one thing about windows. You should add window scripts to UI prefab, not in Startup scene. Just open UI.prefab and see where StartGameWindow is attached.

lexnewgate commented:
OK. I will have a look at it.

Commit:
fix typo and ui prefab

change gtx to gxt

remove window from start scene and add it in prefab

Commit:
add language interface

add language interface;
add some todo comments;
change file tab; make it consistent with project.

lexnewgate commented:
Working on UI Part now. Since I have never used IMGUI, it will take some time.

in0finite commented on Assets/Scripts/Importing/GTX/GXT.cs:
```diff
@@ -7,34 +7,35 @@
+		}
+
+		//this code is pasted from ArchiveMgr; 
+		public static string GameDir => Config.Get<string>("game_dir"); //todo this should be put in a config 
```
GameDir: just use `Config.GamePath` for this

lexnewgate replied:
Yes. Just found it today. Will fix it.

in0finite commented on Assets/Scripts/Importing/GTX/GXT.cs:
```diff
@@ -7,34 +7,35 @@
+
+		//this code is pasted from ArchiveMgr; 
+		public static string GameDir => Config.Get<string>("game_dir"); //todo this should be put in a config 
+		public static string GTXDir => Path.Combine(GameDir, "text");//todo this should be put in a config or just here?
```
no need to put it in config, leave it here

Commit:
fix game path and typo

add test ui

Commit:
add ui part

lexnewgate commented:
![image](https://user-images.githubusercontent.com/9143758/63791147-682cf980-c92d-11e9-871a-2e027a9f0f16.png)

This is a screenshot of GXT window for the moment.

First, click any button in tablename col. Then a col of crc shows. Then click any item in crc. The corresponding text will be shown in the rightmost col.

Commit:
fix set gui style pos

it will cause bug in ctor

lexnewgate commented:
There are some performance issues still. When click the `MAIN` button, the program will freeze a little bit . Because the `MAIN` button contains lots of items. I will work it tomorrow. I know how to handle this issue in UGUI(by using object pool and only show items when they are in viewport), but have no idea how to handle them in IMGUI.

in0finite commented:
Looks good.

To solve performance issues, use pages. Divide content into pages, and let user select current page. See how it's done in peds window.

Commit:
Merge remote-tracking branch 'upstream/dev' into import_GXT

lexnewgate commented:
The page solution solves the performance problem.
![image](https://user-images.githubusercontent.com/9143758/63865479-3b87e900-c9e4-11e9-8512-868855707695.png)
Currently it looks like this. However, when main button clicked.
![image](https://user-images.githubusercontent.com/9143758/63865592-6d00b480-c9e4-11e9-9f1d-aef6a21616db.png)
What do you suggest? Should I make a scroll view of these numbers? Or add two more buttons such as `<<` `>>`. (When clicked, add all indices to a certain number?)

in0finite commented:
Mhm, I understand. I think you should limit number of pages to some number, eg 6. And display buttons `<<` and `>>`, like you said.

Commit:
add pages to gxt ui

lexnewgate commented:
I have make the following function a little messy. 
 ```csharp
public static int DrawPagedViewNumbers (Rect rect, int currentPage, int numPages)
```
I don't know whether should I refactor it by some design patterns? Or just leave it be. Since it may make the function more complex.

in0finite commented:
It seems complicated. Leave paging to me. Can you revert changes to `GUIUtils` ?

And that should be all regarding gxt for now.

Also, revert changes to Startup scene.

in0finite commented:
Actually, if it works fine with both peds and gxt window, then don't revert `GUIUtils`.

lexnewgate commented:
yeah. It works well with both peds and gxt window. So I will only revert startup scene. I am not familiar git yet. I will search how to do it.

lexnewgate commented:
I have reverted the changes to startup scene.

Commit:
revert changes to startup scene
